payments-core/src/main/java/com/stripe/android/view/CardBrandViewRebuild.kt

```diff
@@ -0,0 +1,293 @@
+package com.stripe.android.view
+
+import android.content.Context
+import android.graphics.PorterDuff
+import android.graphics.PorterDuffColorFilter
+import android.graphics.Typeface
+import android.os.Parcelable
+import android.util.AttributeSet
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import android.widget.ArrayAdapter
+import android.widget.FrameLayout
+import android.widget.ImageView
+import android.widget.ListPopupWindow
+import android.widget.TextView
+import com.stripe.android.R
+import com.stripe.android.databinding.StripeCardBrandViewRebuildBinding
+import com.stripe.android.model.CardBrand
+import com.stripe.android.model.CardBrand.Unknown
+import com.stripe.android.model.PaymentMethodCreateParams
+import kotlinx.parcelize.Parcelize
+import kotlin.properties.Delegates
+
+internal class CardBrandViewRebuild @JvmOverloads constructor(
+    context: Context,
+    attrs: AttributeSet? = null,
+    defStyleAttr: Int = 0
+) : FrameLayout(context, attrs, defStyleAttr) {
+    private val viewBinding: StripeCardBrandViewRebuildBinding =
+        StripeCardBrandViewRebuildBinding.inflate(LayoutInflater.from(context), this)
+    private val iconView = viewBinding.icon
+    private val progressView = viewBinding.progress
+    private val chevron = viewBinding.chevron
+    private val listPopup = ListPopupWindow(context)
+
+    @Parcelize
+    internal data class State(
+        val isCbcEligible: Boolean = false,
+        val isLoading: Boolean = false,
+        val brand: CardBrand = Unknown,
+        val userSelectedBrand: CardBrand? = null,
+        val possibleBrands: List<CardBrand> = emptyList(),
+        val merchantPreferredNetworks: List<CardBrand> = emptyList(),
+        val shouldShowCvc: Boolean = false,
+        val shouldShowErrorIcon: Boolean = false,
+        val tintColor: Int = 0,
+    ) : Parcelable
+
+    private var state: State = State()
+
+    var isCbcEligible: Boolean
+        get() = state.isCbcEligible
+        set(value) {
+            state = state.copy(isCbcEligible = value)
+            updateBrandSpinner()
+        }
+
+    var isLoading: Boolean by Delegates.observable(
+        false
+    ) { _, wasLoading, isLoading ->
+        state = state.copy(isLoading = isLoading)
+        setCardBrandIconAndTint()
+        if (wasLoading != isLoading) {
+            if (isLoading) {
+                progressView.visibility = VISIBLE
+            } else {
+                progressView.visibility = GONE
+            }
+        }
+    }
+
+    var brand: CardBrand
+        get() = state.brand
+        set(value) {
+            state = state.copy(brand = value)
```

Comment:
it looks like we're mostly updating one field of `state` at a time. Should we instead keep each of the fields as separate variables? And then when we write the state to savedState, we can convert it to an actual `State` object then?

Reply:
I think this would simplify this class quite a bit

Reply:
@tillh-stripe Wasn't there a particular reason for state restoration when you did this when initially building Card Brand Choice?